### PR TITLE
Validate uniqueness of fixture methods per specification class

### LINKED ISSUE #1512: Count of invocation arguments varies for fixture methods that use default values
### Describe the bug

While writing an extension to the Spock Framework I came across a strange case with the number of arguments I get from the `IMethodInvocation.getArguments()` method in the interceptor implementation. This number depends on the existence of the `cleanup()` method in the specification, the form of Data Tables and the Java version.

See attached [spock-ext.tar.gz](https://github.com/spockframework/spock/files/9460953/spock-ext.tar.gz) and cases described in "Actual behavior".

### To Reproduce

I am including an attachment, with the implementation of the specification, reduced to a minimum, to demonstrate the problem.
[spock-ext.tar.gz](https://github.com/spockframework/spock/files/9460953/spock-ext.tar.gz)

See "Actual behavior" for commands to reproduce this issue.

### Expected behavior

I should get the same results on the standard output, regardless of the above cases.

```
# spock.WithCleanupSpec
method: private java.lang.Object spock.WithCleanupSpec.setup(java.lang.String,java.lang.Integer,java.lang.String)
arguments count: 3
stringValue: overwritten
intValue: 3333
anotherStringValue: overwritten
# spock.WithoutCleanupSpec
method: private java.lang.Object spock.WithoutCleanupSpec.setup(java.lang.String,java.lang.Integer,java.lang.String)
arguments count: 3
stringValue: overwritten
intValue: 3333
anotherStringValue: overwritten
```

### Actual behavior

The Java version seems to make a difference. I ran the test as is in the attached [spock-ext.tar.gz](https://github.com/spockframework/spock/files/9460953/spock-ext.tar.gz) file in the Docker Images listed below, using the following command (where `openjdk:8` is one of the images below):

```console
$ pwd
/tmp/spock-ext
$ docker run --rm -it -w `pwd` -v `pwd`:`pwd` openjdk:8 ./gradlew clean test
```

1. same, but invalid results with these images: `openjdk:8`, `openjdk:10` (not always)

```
# spock.WithCleanupSpec
method: private java.lang.Object spock.WithCleanupSpec.setup(java.lang.String,java.lang.Integer)
arguments count: 2
stringValue: overwritten
intValue: 3333
anotherStringValue: defaultValue
# spock.WithoutCleanupSpec
method: private java.lang.Object spock.WithoutCleanupSpec.setup(java.lang.String,java.lang.Integer)
arguments count: 2
stringValue: overwritten
intValue: 3333
anotherStringValue: defaultValue
```

2. different results with these images: `openjdk:9`, `openjdk:10` (only once, after switching from `openjdk:9`)
```
# spock.WithCleanupSpec
method: private java.lang.Object spock.WithCleanupSpec.setup(java.lang.String,java.lang.Integer,java.lang.String)
arguments count: 3
stringValue: overwritten
intValue: 3333
anotherStringValue: overwritten
# spock.WithoutCleanupSpec
method: private java.lang.Object spock.WithoutCleanupSpec.setup(java.lang.String,java.lang.Integer)
arguments count: 2
stringValue: overwritten
intValue: 3333
anotherStringValue: defaultValue
```

3. another different results with these images: `openjdk:10` (not always), `openjdk:11`, `openjdk:12`, `openjdk:13`, `openjdk:14`
```
# spock.WithCleanupSpec
method: private java.lang.Object spock.WithCleanupSpec.setup(java.lang.String,java.lang.Integer)
arguments count: 2
stringValue: overwritten
intValue: 3333
anotherStringValue: defaultValue
# spock.WithoutCleanupSpec
method: private java.lang.Object spock.WithoutCleanupSpec.setup(java.lang.String)
arguments count: 1
stringValue: overwritten
intValue: 2222
anotherStringValue: defaultValue
```

4. another same, but still invalid results with these images (once, after switching the Java version): `openjdk:11`, `openjdk:12`
```
# spock.WithCleanupSpec
method: private java.lang.Object spock.WithCleanupSpec.setup(java.lang.String)
arguments count: 1
stringValue: overwritten
intValue: 2222
anotherStringValue: defaultValue
# spock.WithoutCleanupSpec
method: private java.lang.Object spock.WithoutCleanupSpec.setup(java.lang.String)
arguments count: 1
stringValue: overwritten
intValue: 2222
anotherStringValue: defaultValue
```

### Java version

Java version depending on the Docker image used.

```console
$ docker run --rm -it openjdk:8 java -version 
openjdk version "1.8.0_342"
OpenJDK Runtime Environment (build 1.8.0_342-b07)
OpenJDK 64-Bit Server VM (build 25.342-b07, mixed mode)

$ docker run --rm -it openjdk:9 java -version
openjdk version "9.0.4"
OpenJDK Runtime Environment (build 9.0.4+12-Debian-4)
OpenJDK 64-Bit Server VM (build 9.0.4+12-Debian-4, mixed mode)

$ docker run --rm -it openjdk:10 java -version
openjdk version "10.0.2" 2018-07-17
OpenJDK Runtime Environment (build 10.0.2+13-Debian-2)
OpenJDK 64-Bit Server VM (build 10.0.2+13-Debian-2, mixed mode)

$ docker run --rm -it openjdk:11 java -version
openjdk version "11.0.16" 2022-07-19
OpenJDK Runtime Environment 18.9 (build 11.0.16+8)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.16+8, mixed mode, sharing)

$ docker run --rm -it openjdk:12 java -version
openjdk version "12.0.2" 2019-07-16
OpenJDK Runtime Environment (build 12.0.2+10)
OpenJDK 64-Bit Server VM (build 12.0.2+10, mixed mode, sharing)

$ docker run --rm -it openjdk:13 java -version
openjdk version "13.0.2" 2020-01-14
OpenJDK Runtime Environment (build 13.0.2+8)
OpenJDK 64-Bit Server VM (build 13.0.2+8, mixed mode, sharing)

$ docker run --rm -it openjdk:14 java -version
openjdk version "14.0.2" 2020-07-14
OpenJDK Runtime Environment (build 14.0.2+12-46)
OpenJDK 64-Bit Server VM (build 14.0.2+12-46, mixed mode, sharing)
```

### Buildtool version

For Image `openjdk:14`:

```console
$ ./gradlew --version

------------------------------------------------------------
Gradle 7.4.2
------------------------------------------------------------

Build time:   2022-03-31 15:25:29 UTC
Revision:     540473b8118064efcc264694cbcaa4b677f61041

Kotlin:       1.5.31
Groovy:       3.0.9
Ant:          Apache Ant(TM) version 1.10.11 compiled on July 10 2021
JVM:          14.0.2 (Oracle Corporation 14.0.2+12-46)
OS:           Linux 5.15.0-46-generic amd64
```

### What operating system are you using

Linux

### Dependencies

[dependencies-report.txt](https://github.com/spockframework/spock/files/9461060/dependencies-report.txt)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by leonard84**
fixes #1512

### CODE REVIEW

File: docs/spock_primer.adoc [hunk 1]
```diff
@@ -81,7 +81,7 @@ the field at the very beginning of the `setupSpec()` method.)
 static final PI = 3.141592654
 ----
 
-Static fields should only be used for constants. Otherwise shared fields are preferable, because their semantics with
+Static fields should only be used for constants.Otherwise shared fields are preferable, because their semantics with
```
Comment (szpak) [inline docs/spock_primer.adoc:84]: An accidental change?

**Review by szpak — APPROVED** [review]

File: docs/spock_primer.adoc (hunk 1 above)
Comment (leonard84) [reply docs/spock_primer.adoc:84]: Weird, must be some intellij autoformat going on. I only added L105+

File: spock-specs/src/test/groovy/org/spockframework/runtime/model/SpecInfoSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  def $fixtureMethod(String param = 'default') {
+    println param
+  }
+
+  def test() {
+    expect:
+    true
+  }
+}
+""")
```
Comment (piotrminkina) [inline spock-specs/src/test/groovy/org/spockframework/runtime/model/SpecInfoSpec.groovy:24]: Try this formating style:

```groovy
    def classes = new EmbeddedSpecCompiler().compile("""\
      import spock.lang.Specification
      class ASpec extends Specification {
        def $fixtureMethod(String param = 'default') {
          println param
        }
        def test() {
          expect:
          true
        }
      }
      """.stripIndent())
```

or call `.stripIndent()` always in `.compile()` method, or define new method `.compileIndented()`.

**Review by piotrminkina — APPROVED** [review]
Great! Thank you very much :)

File: spock-specs/src/test/groovy/org/spockframework/runtime/model/SpecInfoSpec.groovy (hunk 2 above)
Comment (leonard84) [reply spock-specs/src/test/groovy/org/spockframework/runtime/model/SpecInfoSpec.groovy]: White space for the embedded code is irrelevant, but I've changed the indentation

**Pull request merged**